docs/rules/require-reduce-negation.md

```diff
@@ -0,0 +1,92 @@
+---
+pageClass: "rule-details"
+sidebarDepth: 0
+title: "regexp/require-reduce-negation"
+description: "require to reduce negation of character classes"
+---
+# regexp/require-reduce-negation
+
+💼 This rule is enabled in the ✅ `plugin:regexp/recommended` config.
+
+🔧 This rule is automatically fixable by the [`--fix` CLI option](https://eslint.org/docs/latest/user-guide/command-line-interface#--fix).
+
+<!-- end auto-generated rule header -->
+
+> require to reduce negation of character classes
+
+## :book: Rule Details
+
+This rule is aimed at optimizing patterns by reducing the negation (complement) representation of character classes (with `v` flag).
+
+<eslint-code-block fix>
+
+```js
+/* eslint regexp/require-reduce-negation: "error" */
+
+/* ✗ BAD */
+var re = /[^[^abc]]/v; // -> /[abc]/v
+var re = /[^\D]/u; // -> /[\d]/u
```

Comment:
So this rule is a superset of [`regexp/negation`](https://ota-meshi.github.io/eslint-plugin-regexp/rules/negation.html)? I'm not quite sure what to do here. I think that it makes sense for `regexp/require-reduce-negation` to fix this issue, but then why does ``regexp/negation`` even exist?

I think we have a few choices here:

1. Just let it be. We'll leave both rules as is and think of `regexp/negation` as a simpler `regexp/require-reduce-negation`.
2. We remove `regexp/negation`.
3. We restrict `regexp/require-reduce-negation` to only fix issues that `regexp/negation` does not fix.

Reply:
Hmm..  I think the original purpose of the `negation` rule is to rewrite a pattern that everyone considers meaningless into a correct pattern. For example, in JS, we would like to prohibit things like `if(!true)` and `if(!false)`.
I think the new rule should be main purpose at pattern optimization.
Therefore, in my opinion, I would like to make the two rules separate.

I think that the pattern `/[^[^abc]]/v;` might also be reported in the `negation` rule, but I'm not sure if new rule need to ignore it. Because it is reported both for optimization and for finding mistakes.

Reply:
I like this idea.

As for `negation`, I think we can justify the new functionality by saying that `negation` applies the negation to the character class element. So for some `[^x]`, it will always try to negate `x` (no matter what that `x` is). This would justify `[^[a]]` -> `[^a]`, `[^[^a]]` -> `[a]`, `[^\s]` -> `[\S]` -> `\S`, and `[^[a&&b]]` -> `[^a&&b]`.

As for `De Morgan` and `Subtraction to intersection, and intersection to subtraction`, they might have to be one rule. The issue is that `Subtraction to intersection, and intersection to subtraction` might want to convert `[^a]&&[^b]` to `[^a]--b`. So maybe the combination of the two could be this rule?

Reply:
> The issue is that Subtraction to intersection, and intersection to subtraction might want to convert [^a]&&[^b] to [^a]--b. So maybe the combination of the two could be this rule?

Ah, that's true. If we add two rules, the rules may conflict. It makes sense to add one rule. I have no idea for the name of the new rule. Do you have any ideas for good rule name?

Reply:
Maybe something like `simplify-set-operations`?

Reply:
Thanks for sharing your idea! I'll rename the rule to that name later.

Reply:
I changed the rule name and changed the docs.